runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java

```diff
@@ -58,6 +86,107 @@ public SamzaRunner(SamzaPipelineOptions options) {
     this.options = options;
   }
 
+  private void closeFnServer(GrpcFnServer<?> fnServer) {
+    try (AutoCloseable closer = fnServer) {
+      // do nothing
+    } catch (Exception e) {
+      LOG.error("Failed to close fn api servers. Ignore since this is shutdown process...", e);
+    }
+  }
+
+  private void setUpContextManager(
+      StreamGraph streamGraph, SamzaExecutionContext executionContext) {
+    streamGraph.withContextManager(
+        new ContextManager() {
+          @Override
+          public void init(Config config, TaskContext context) {
+            if (executionContext.getMetricsContainer() == null) {
+              final MetricsRegistryMap metricsRegistry =
+                  (MetricsRegistryMap) context.getSamzaContainerContext().metricsRegistry;
+              executionContext.setMetricsContainer(new SamzaMetricsContainer(metricsRegistry));
+            }
+
+            if (SamzaRunnerOverrideConfigs.isPortableMode(options)) {
+              if (jobBundleFactory == null) {
```

Comment:
Hmm, this would tie to the work to test "multiple sdk workers connecting to one java runner" as well as replacing "SingleEnvironmentInstanceJobBundleFactory". I think it's better to group those effort together